src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegion.java

```diff
@@ -0,0 +1,609 @@
+        Utils.validateArg(minRegionSize <= maxRegionSize, "minRegionSize must be <= maxRegionSize");
+        Utils.validateArg(assemblyRegionPadding >= 0, "assemblyRegionPadding must be >= 0");
+        Utils.validateArg(activeProbThreshold >= 0.0, "activeProbThreshold must be >= 0.0");
+        Utils.validateArg(maxProbPropagationDistance >= 0, "maxProbPropagationDistance must be >= 0");
+
+        // TODO: refactor this method so that we don't need to load all reads from the shard into memory at once!
+        final List<GATKRead> windowReads = new ArrayList<>();
+        for ( final GATKRead read : shard ) {
+            windowReads.add(read);
+        }
+
+        final LocusIteratorByState locusIterator = new LocusIteratorByState(windowReads.iterator(), DownsamplingMethod.NONE, false, false, ReadUtils.getSamplesFromHeader(readsHeader), readsHeader);
+        final ActivityProfile activityProfile = new BandPassActivityProfile(null, maxProbPropagationDistance, activeProbThreshold, BandPassActivityProfile.MAX_FILTER_SIZE, BandPassActivityProfile.DEFAULT_SIGMA, readsHeader);
+
+        // First, use our activity profile to determine the bounds of each assembly region:
+        List<AssemblyRegion> assemblyRegions = new ArrayList<>();
+        for ( final AlignmentContext pileup : locusIterator ) {
+            if ( ! activityProfile.isEmpty() ) {
+                final boolean forceConversion = pileup.getLocation().getStart() != activityProfile.getEnd() + 1;
+                assemblyRegions.addAll(activityProfile.popReadyAssemblyRegions(assemblyRegionPadding, minRegionSize, maxRegionSize, forceConversion));
+            }
+
+            if ( shard.getPaddedInterval().contains(pileup.getLocation()) ) {
+                final SimpleInterval pileupInterval = new SimpleInterval(pileup.getLocation());
+                final ReferenceBases refBase = new ReferenceBases(new byte[]{referenceContext.getBases()[pileup.getLocation().getStart() - referenceContext.getWindow().getStart()]}, pileupInterval);
+                final ReferenceContext pileupRefContext = new ReferenceContext(new ReferenceMemorySource(refBase, readsHeader.getSequenceDictionary()), pileupInterval);
+
+                final ActivityProfileState profile = evaluator.isActive(pileup, pileupRefContext, features);
+                activityProfile.add(profile);
+            }
+        }
+        assemblyRegions.addAll(activityProfile.popReadyAssemblyRegions(assemblyRegionPadding, minRegionSize, maxRegionSize, true));
+
+        // Then, fill the assembly regions with overlapping reads from the shard:
+        final PeekableIterator<GATKRead> reads = new PeekableIterator<>(windowReads.iterator());
+        AssemblyRegion lastRegion = null;
+        for ( final AssemblyRegion region : assemblyRegions ) {
+            if ( lastRegion != null ) {
+                for ( final GATKRead lastRegionRead : lastRegion.getReads() ) {
```

Comment:
I think this is a rare operation not deserving of a new method + tests-- we only do it here because reads can span multiple regions, so we need to check the reads in the previous region for overlap with the current region.